### PR TITLE
UnitY2/ProsodyUnitY2 model training

### PR DESCRIPTION
**Pull request by yilinyang7**
This is an WIP version of UnitY2/ProsodyUnitY2 training codebase.
It includes (I'll write more details in a seperate doc):
* loss computation (at unity/loader.py)
* dataset class written 100% with fairseq2.data API (unity2/dataset.py)
* trainer class & pipeline (unity2/trainer.py)
* implemented trainable UnitYAlignerModel, UnitYNART2Umodel, UnitYModel, etc..
* yaml-configurable multi-task learning model (multi_task_configs.py)
* cli-configurable multi-task learning objective (unity2/trainer.py)

Still WIP part:
* multi-node job launcher
* UnitYMetricBag class (low-pri)

There're also numerious backward-compatible changes.

### CODE REVIEW

File: src/seamless_communication/cards/datasets/seamless_expressive_eng_cmn.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        train_mexpresso_clean_src_eng_cmn: 10239
+        train_dubbed_clean_src_cmn_eng: 228872
+        train_dubbed_clean_src_eng_cmn: 228823
+        train_mined_clean_src_mg1_06_cmp3_2_cmn_eng: 57914
+        train_mined_clean_src_mg1_06_cmp3_2_eng_cmn: 57865
+        train_tmodule_clean_src_cos_0_75_cmp_3_2_cmn_eng: 52966
+        train_tmodule_clean_src_cos_0_75_cmp_3_2_eng_cmn: 52971
+    valid:
+        dev_mexpresso_dubbed_clean_src_cmn_eng: 3459
+        dev_mexpresso_dubbed_clean_src_eng_cmn: 3445
```
Comment (pipibjc) [inline src/seamless_communication/cards/datasets/seamless_expressive_eng_cmn.yaml:25]: nit: these can't be exposed to public repo due to 1) private path, 2) expose more details of data source (dubbed, mined, etc)

File: src/seamless_communication/cards/seamless_expressivity_trainable.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# MIT_LICENSE file in the root directory of this source tree.
+
+name: seamless_expressivity_trainable
+base: unity_nllb-100
+model_arch: expressivity_v2
+char_tokenizer: "https://huggingface.co/facebook/seamless-streaming/resolve/main/spm_char_lang38_tc.model"
+checkpoint: "/fsx-ust/shared/seamless23_oss/m2m_expressive_unity.pt"
```
Comment (pipibjc) [inline src/seamless_communication/cards/seamless_expressivity_trainable.yaml:11]: nit: fine with internal, but maybe replace with placeholder when landing in public repo

**Review by pipibjc — COMMENTED** [review]
thanks Yilin!!! I left some very shallow comments. I will take a deeper look.

File: src/seamless_communication/cards/datasets/seamless_expressive_eng_cmn.yaml (hunk 1 above)
Comment (yilinyang7) [reply src/seamless_communication/cards/datasets/seamless_expressive_eng_cmn.yaml:25]: I see, this is required for model training (i.e. the trainer would load the dataset from this yaml file), this is how fairseq2 designs.

The number of rows (e.g. 866399) weren't required actually, and I could remove them.
But I agree the path should be hidden for private purposes. 

In that case, I could just upload a dummy dataset for open-sourcing.

File: src/seamless_communication/cards/seamless_expressivity_trainable.yaml (hunk 2 above)
Comment (yilinyang7) [reply src/seamless_communication/cards/seamless_expressivity_trainable.yaml:11]: Yes, sure.

**Pull request closed without merging**